apm-agent-plugins/apm-jaxrs-plugin/src/main/java/co/elastic/apm/jaxrs/JaxRsTransactionNameInstrumentation.java

```diff
@@ -0,0 +1,90 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.jaxrs;
+
+import co.elastic.apm.bci.ElasticApmInstrumentation;
+import co.elastic.apm.bci.bytebuddy.SimpleMethodSignatureOffsetMappingFactory.SimpleMethodSignature;
+import co.elastic.apm.impl.ElasticApmTracer;
+import co.elastic.apm.impl.stacktrace.StacktraceConfiguration;
+import co.elastic.apm.impl.transaction.Transaction;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.matcher.ElementMatchers;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+
+import static co.elastic.apm.bci.bytebuddy.CustomElementMatchers.classLoaderCanLoadClass;
+import static co.elastic.apm.bci.bytebuddy.CustomElementMatchers.isInAnyPackage;
+import static net.bytebuddy.matcher.ElementMatchers.isAnnotatedWith;
+import static net.bytebuddy.matcher.ElementMatchers.isBootstrapClassLoader;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+
+public class JaxRsTransactionNameInstrumentation extends ElasticApmInstrumentation {
+
+    private Collection<String> applicationPackages = Collections.emptyList();
+
+    @Advice.OnMethodEnter
+    private static void setTransactionName(@SimpleMethodSignature String signature) {
+        if (tracer != null) {
+            final Transaction transaction = tracer.currentTransaction();
+            if (transaction != null) {
+                transaction.withName(signature);
+            }
+        }
+    }
+
+    @Override
+    public void init(ElasticApmTracer tracer) {
+        applicationPackages = tracer.getConfig(StacktraceConfiguration.class).getApplicationPackages();
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        // setting application_packages makes this matcher more performant but is not required
+        return isInAnyPackage(applicationPackages, ElementMatchers.<TypeDescription>any())
```

Comment:
Don't you think it is likely that there are frameworks that use JAX-RS? I am not sure what we want to do if there are, but this means we will miss them, right?

Reply:
Good point. Not sure if that is a likely case. The use case would be when importing a 3rd party library which also contains JAX-RS resources, right?

Reply:
Yes, that what I meant

Reply:
I am not sure if this check provides any value...

Reply:
Quite the contrary. This `isInAnyPackage` matcher is a really cheap one because the class file does not have to be parsed at all. In general, the matchers which match based on the class names are the cheapest ones. Without this check, each and every class has to be parsed to see if it is annotated with `@Path`. Prepending the `isInAnyPackage` matcher considerably limits the amount classes which have to be checked with a more expensive matcher :). Also, typically, `application_packages` is only set to one or a handful of packages.

In the case you are importing a 3rd party library containing JAX-RS resources, it probably makes sense to add that to the application_packages namespace anyway. Because then stack traces also highlight those classes. But I can see that this may be confusing for users which wonder why some endpoints are not picked up. It's a tradeoff but I think with proper documentation (right at the supported technologies page), this would be the tradeoff I prefer.

Reply:
> Without this check, each and every class has to be parsed to see if it is annotated with `@Path`. Prepending the `isInAnyPackage` matcher considerably limits the amount classes which have to be checked with a more expensive matcher :)

Right, my mistake, I confused with the CL matcher that use caches...